src/main/java/tree/ImpTargetTreeNode.kt

```diff
@@ -1,30 +1,100 @@
 package tree
 
-class ImpTargetTreeNode : TargetTreeNode {
-    override fun moveRight(): TargetTreeNode {
-        println("moveRight")
-        return this
+class ImpTargetTreeNode() : TargetTreeNode {
+
+    private var name: String = ""
+    private var tag: String = ""
+    private var id: Int = -1
+    private var children: List<TargetTreeNode>? = null
+    private var parent: TargetTreeNode? = null
+
+    constructor(name: String, tag: String, id: Int) : this() {
+        this.name = name
+        this.tag = tag
+        this.id = id
+    }
+
```

Comment:
If we do not need the default `constructor` to this class (`val targetTree = ImpTargetTreeNode()`) we can compress this for:
```suggestion
class ImpTargetTreeNode(val name: String, val tag: String, val id: Int) : TargetTreeNode {
    private var children: List<TargetTreeNode>? = null
    private var parent: TargetTreeNode? = null
```
The class properties don't seem to change so maybe we can use `val`. 
![](https://media.giphy.com/media/DZu3mnCcmsJKE/giphy.gif)

Reply:
We are using the default constructor, so I don't think that it might be possible. However, I think the `var` problem might be addressed. Even more so with some of the changes proposed earlier.

Reply:
But what about this change? We could get rid of the constructor and 3 `var`.

```
class ImpTargetTreeNode(
        private val name: String = "",
        private val tag: String = "",
        private var id: Int = -1
    ) : TargetTreeNode {

    private var children: List<TargetTreeNode>? = null
    private var parent: TargetTreeNode? = null
```

Reply:
So, combining two different requests, the agreed solution by the team was moving some of the propeties to the interface and leaving them as val.